hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala

```diff
@@ -317,25 +317,25 @@ abstract class PContainer extends PIterable {
       return Code._empty
     }
 
+    val missingHeaderOffset = sourceOffset + sourceType.lengthHeaderBytes
+    val n = PContainer.loadLength(sourceOffset).toL
+    val m1 = (n.toI / 32).toL
     val i = mb.newLocal[Long]
     Code(
-      i := PContainer.nMissingBytes(sourceType.loadLength(sourceOffset)).toL,
-      Code.whileLoop(i > 0L,
-        (i >= 8L).mux(
-          Code(
-            i := i - 8L,
-            Region
-              .loadLong(sourceOffset + sourceType.lengthHeaderBytes + i)
-              .cne(const(0.toByte))
-              .orEmpty(onFail)
-          ),
-          Code(
-            i := i - 1L,
-            Region
-              .loadByte(sourceOffset + sourceType.lengthHeaderBytes + i)
-              .cne(const(0.toByte))
-              .orEmpty(onFail)
-          )
+      i := 0L,
+      Code.whileLoop(
+        i < m1,
+        Code(
+          Region.loadInt(missingHeaderOffset + i * const(4L)).cne(const(0)).orEmpty(onFail),
+          i := i + const(1L)
+        )
+      ),
+      i := i * 32L,
```

Comment:
removed the middle byte loop because there is no substantial benefit to having it, maximum of 3 bytes read if something isn't divisible by 4.